### ISSUE TITLE
add ability to run tasks in parallel, and to set the level of parallelism via env var or in .looprc

### ISSUE THREAD

**Issue by mateodelnorte** [mateodelnorte/loop#16]

**Comment by shawninder** [comment]
This sound especially interesting for things like `loop "npm i"` which can be quite long if run sequentially

**Comment by mateodelnorte** [comment]
Yup. Agreed. Unfortunately, this is all on hold until I'm done with a token launch.

**Comment by patrickleet** [comment]
I've started running my whole dockerized CI process on every push, and it takes a while - running in parallel would be huge

**Comment by mateodelnorte** [comment]
The only problem is for stdout and stderr. Since they are now concurrent
processes, a decision needs to be made about how to log. Giant mess?
Redirect to files?

**Comment by shawninder** [comment]
I don't know what the best choice is here, but just to add another option for consideration, you could cache the logs and output them as if they had run sequentially at the end...

**Comment by patrickleet** [comment]
docker-compose outputs container logs as a giant mess that has color coded labels like debug, and an application, kitematic to see the logs as a not jumbled mess. What about piggy backing, and kind of avoiding the issue by running everything via docker?

```
loop "npm i" --image=node:8-alpine
```

```
const docker = require('dockerode')
if (image)
  docker.run(image, command.split, process.stdout, function (err, data, container) {
    console.log(data.StatusCode);
  });
}
```

Then all the logs would be separate in Kitematic (in theory - haven't tried any of this)

**Comment by patrickleet** [comment]
redirect to files sounds simplest

**Comment by mateodelnorte** [comment]
People shouldn't have to have docker either. But I could easily create
known files in a temp folder.

**Comment by gabemeola** [comment]
Giant mess with with filter tool? Thinking of something like [`jest --watch` mode](https://github.com/mateodelnorte/loop/issues/16)

**Comment by gabemeola** [comment]
Was reply too your comment https://github.com/mateodelnorte/loop/issues/16#issuecomment-341452471 . I thinking more of long running processes such as multiple servers.

**Comment by mateodelnorte** [comment]
Ah. Hah. Been so long since the comment. Tracking now. 🤓

**Comment by LarsOL** [comment]
This would be useful, especially for your other tool 'meta', when we run `meta git pull` it can take a long time. Would be helpful if it was parralized .

Jest deals with the console by [buffering](https://facebook.github.io/jest/blog/2016/09/01/jest-15.html#buffered-console-messages). Under the hood they use a [BufferedConsole](https://github.com/facebook/jest/blob/master/packages/jest-runner/src/run_test.js) maybe it could be implemented similarly?

**Comment by johannessjoberg** [comment]
Any updates concerning this?

**Comment by bburgin** [comment]
Could meta call this to parallelize commands?:
https://github.com/shenwei356/rush

**Comment by mateodelnorte** [comment]
This has been implemented.

Want to give it a shot?

**Comment by mateodelnorte** [comment]
This issue was fixed in https://github.com/mateodelnorte/loop/commit/79cdc88ccc1c958ea36abe6f3095c50f23f12e84

**Comment by mateodelnorte** [comment]
If anyone is interested in this feature, please test it out. Thanks
